doc/UnifiedTeamD/Unified_TeamD_HLD.md

```diff
@@ -0,0 +1,426 @@
+
+Once initialized, teamd can dynamically manage port channels based on configuration updates from teammgrd.
+
+**Docker Container restart**
+
+When the system starts, teamd operates in the new design where it runs in single-process mode and manages all PortChannels through IPC. To downgrade to the legacy design, the mode must be explicitly set to "multi-process" and the Docker container must be restarted. Similarly, to upgrade back to the new design, the "multi-process" mode configuration should be removed, followed by a Docker restart. Therefore, the mode setting must be correctly configured before restarting the container to ensure the system comes up in the desired operation mode.
+
+Refer section [teamd operating mode](#teamd-operating-mode) for more details on how to set the mode.
+
+## Libteam
+
+Libteam is a user-space library that provides APIs for creating, configuring, and managing team devices while interfacing with the kernel through netlink sockets. It also applies the runner logic to control link aggregation behavior in the kernel.
+
+**Introduction of team_netlink Structure**
+
+A new structure "team_netlink" will be defined in the libteam layer to hold all shared netlink socket details. This structure enables reuse of common socket connections across multiple team devices, avoiding the creation of individual netlink sockets for each instance.
+
+**Optimized Netlink Communication**
+
+A shared set of netlink sockets will be created once and used for communication with the kernel. These sockets will handle team-related control messages and events for all team (portchannel) devices, by maintaining a map between ifindex to corresponding portchannels.
+
+**Netlink buffer size**
+
+As multiple LAGs are handled by single teamd daemon in case of unified-mode, the netlink buffer size is increased from current 960kB to 3MB to handle the increased number of netlink messages.
+
+## FD Optimization and Wheel Timers 
+
+Currently each LAG uses 1 timerfd and 2 fds for pipe, each member port uses 3 timer fds and 1 fd for socket used for packet processing. The pipe is used for internal communication while doing LACP port aggregator selection and for active port selection with active-backup runner. With single unified teamd process, as there will be multiple LAGs in single teamd process this can lead to hitting the FD limit of 1024 for the process when scaled. To address this, following changes are made:
+
+- Instead of one pipe for each LAG, a single pipe will be used in global context and used across all LAGs 
+- 2 timer fds will be created, one for timers with granularity of seconds and another for 100ms granularity. These timers will then be used to add support for existing timers in teamd using hashed wheel timer. 
+
+### Overview: What Is a Wheel Timer? 
+
+A wheel timer organizes timers in a circular buffer structure where: 
+
+- Each slot in the "wheel" represents a time interval (e.g., 100ms, 1sec). 
```

Comment:
Check if 100ms time interval is required, if not, just use the 1 sec time interval.